### DISCUSSION TITLE
[help] Understanding branching

### DISCUSSION THREAD

**Discussion by llrs** [ropensci/targets#1253] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I have/had a lot of trouble grasping how branching works.

When I went to run the [static branching example](https://books.ropensci.org/targets/static.html#map) from the book, I found out it don't work.
The problem, as far as I understood it, is that `method1`, `method2` and `summarize_analysis` do not exists, as such I created a full working example with those functions. 

I have come up with this, that could be added in the book if it helps:

```r
# _targets.R file:
library(targets)
library(tarchetypes)
library(tibble)

values <- expand_grid( # Use all possible combinations of input settings.
  method_function = rlang::syms(c("method1", "method2")),
  data_source = list(c("NIH", "NA"), "NIAID"),
)
values$name <- paste0(values$method_function, "_", c("A", "B"))

static <- tar_map(
  values = values,
  names = "name", # Select columns from `values` for target names.
  tar_target(analysis, method_function(data_source, reps = 10)),
  tar_target(summary, summarize_analysis(analysis))
)

# functions and methods used via NSE
method1 <- function(...){paste0(..., collapse = "")}
method2 <- paste
summarize_analysis <- length
list(static)
```

----

About the dynamic branching, in the `tar_group` examples it is unclear to me why it is needed the `iteration = "group"` (+`tar_group`) to be able to later use `iteration = "vector"`. 
How could one use `iteration = "vector"` without a previous target with `iteration = "group"`?

----

In addition, would be possible to have dynamic branching with names?
Here is my idea about how to name the different iterations approaches
 - `iteration = "group"` could collapse the grouping factor with "_" or just use the tar_group column instead of the hash 
 - `iteration = "list"` could use the names of the list (if they exists)
 - `iteration = "vector"` could collapse the grouping factor with "_" or just use the tar_group column instead of the hash. 

This would help with to interpret the resulting object. This idea is similar to #672, specially [this comment](https://github.com/ropensci/targets/discussions/668#discussioncomment-1518935)

**Comment by wlandau** [comment]
> The problem, as far as I understood it, is that method1, method2 and summarize_analysis do not exists, as such I created a full working example with those functions.

I suppose it would be easier if those functions were defined. We could just as easily use functions in `base`. A PR would help.

> About the dynamic branching, in the tar_group examples it is unclear to me why it is needed the iteration = "group" (+tar_group) to be able to later use iteration = "vector".

`iteration = "group"` prepares a data frame target to be branched over in a downstream target, whereas `iteration = "vector"` applies controls how the return values are formatted in that downstream target.

> How could one use iteration = "vector" without a previous target with iteration = "group"?

`iteration = "group"` only applies to branching over row groups. Without `iteration = "group"` in the upstream target, any downstream targets with `iteration = "vector"` will branch over the natural `vctrs::vec_slice()` slices of the upstream target.

> In addition, would be possible to have dynamic branching with names?

The values returned in branches could have any number of forms, from vectors to multi-row data frames to environments to `R6` objects. It is not be reliable in the general case to extract nice names from these kinds of objects. And if the names were separately  specified by the user, the user would always have to manually check that they line up with the branches, which is not reliable because the number of branches is not generally known in advance. So it would not be feasible to try implementing custom dynamic branch names in `targets`.

> or just use the tar_group column instead of the hash

The `tar_group` column is an ordered vector of integer indexes, which are not suitable for dynamic branch names. If you begin with branches 1, 2, 3, and 4, then rerun the pipeline inserting a branch in the middle, then the re-indexing would invalidate half the branches and cause them to rerun. Hashes are determined independently from other branches, and so they allow branches to be skipped even when other branches are inserted. I don't want users to have to understand this concept, I would rather have the current guardrails in place so users cannot make the mistake of using a sequence integers as dynamic branch names.

**Reply by llrs** [reply]
Thanks for the detailed explanation on why such naming wouldn't work or how the name is used for comparing the hashes and the potential problems with introducing a new branch in the middle.
